zerver/lib/user_groups.py

```diff
@@ -241,7 +271,19 @@ def create_system_user_groups_for_realm(realm: Realm) -> Dict[int, UserGroup]:
         everyone_on_internet_system_group,
     ]
 
-    UserGroup.objects.bulk_create(system_user_groups_list)
+    with transaction.atomic():
+        creation_time = timezone_now()
+        UserGroup.objects.bulk_create(system_user_groups_list)
+        RealmAuditLog.objects.bulk_create(
+            RealmAuditLog(
+                realm=realm,
+                acting_user=None,
+                event_type=RealmAuditLog.USER_GROUP_CREATED,
+                event_time=creation_time,
+                modified_user_group=user_group,
+            )
+            for user_group in system_user_groups_list
```

Comment:
While we are at this, do we also want to incldue creating `GroupGroupMembership` objects in the atomic transaction? Though `create_system_groups_for_realm` is called inside a transaction from `do_create_realm` but not when calling from `do_impor_realm`.

Reply:
That makes sense. We don't want to end up in a bad state when the system groups are half-prepared. There are functions in `zerver.lib.user_groups` that I want to move to `zerver.actions.user_groups`. With a separate refactoring, we might wrap this entire function inside a transaction.